homeassistant/components/nut/const.py

```diff
@@ -570,6 +568,240 @@
+        icon="mdi:flash",
+        native_unit_of_measurement=ELECTRIC_CURRENT_AMPERE,
+        state_class=SensorStateClass.MEASUREMENT,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+    ),
+    "outlet.1.delay.shutdown": SensorEntityDescription(
+        key="outlet.1.delay.shutdown",
+        name="Outlet 1 Shutdown Delay",
+        native_unit_of_measurement=TIME_SECONDS,
+        device_class=SensorDeviceClass.DURATION,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+    ),
+    "outlet.1.delay.start": SensorEntityDescription(
+        key="outlet.1.delay.start",
+        name="Outlet 1 Start Delay",
+        native_unit_of_measurement=TIME_SECONDS,
+        device_class=SensorDeviceClass.DURATION,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+    ),
+    "outlet.1.desc": SensorEntityDescription(
+        key="outlet.1.desc",
+        name="Outlet 1 Description",
+        icon="mdi:information-outline",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+    ),
+    "outlet.1.id": SensorEntityDescription(
+        key="outlet.1.id",
+        name="Outlet 1 Identifier",
+        icon="mdi:information-outline",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+    ),
+    "outlet.1.power": SensorEntityDescription(
+        key="outlet.1.power",
+        name="Outlet 1 Apparent Power",
+        icon="mdi:flash",
+        native_unit_of_measurement=POWER_VOLT_AMPERE,
+        state_class=SensorStateClass.MEASUREMENT,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+    ),
+    "outlet.1.powerfactor": SensorEntityDescription(
+        key="outlet.1.powerfactor",
+        name="Outlet 1 Power Factor",
+        icon="mdi:gauge",
+        native_unit_of_measurement=PERCENTAGE,
+        state_class=SensorStateClass.MEASUREMENT,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+    ),
+    "outlet.1.realpower": SensorEntityDescription(
+        key="outlet.1.realpower",
+        name="Outlet 1 Real Power",
+        native_unit_of_measurement=POWER_WATT,
+        device_class=SensorDeviceClass.POWER,
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    "outlet.1.status": SensorEntityDescription(
+        key="outlet.1.status",
+        name="Outlet 1 Status",
+        icon="mdi:information-outline",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+    ),
+    "outlet.1.switchable": SensorEntityDescription(
+        key="outlet.1.switchable",
+        name="Outlet 1 Switchable",
+        icon="mdi:information-outline",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+    ),
```

Comment:
Please generate these dynamically using `dataclasses.replace` based on what the device supports

Reply:
Thanks for taking a look at this Pull Request, it's my first one on the Home Assistant project so I'm hoping you can be patient and understanding with me.


I am happy to do any work that is required to get this working and done properly, but I have no idea where to start when you say that you want me to use `dataclasses.replace`.

I have my NUT output direct from the `upsc` commands or from the Diagnostics File that can be downloaded from the device page via the NUT Integration itself in Home Assistant.  Not sure how what Python code I would need to run to generate the `SENSOR_TYPES` for `const.py` as you requested.


For my testing, I just made the changes to the `const.py` file manually and brought it in to my docker instance to test.  After Restarting Home Assistant, Deleting the current UPS, re-adding it as a new NUT Integration, it was properly reading all the values and reporting them in Home Assistant as valid entities.

I have 3 different model Eaton UPS's that I can generate data for and a handful of APC units as well.

Reply:
I'd make a new constant list for the dynamically generated once and check to see if the value exists by enumerating the number of potential outlets and then call replace on each of the matching ones to generate a new list of descriptions for the device.

Reply:
Okay, I'll have to take a look at `dataclasses `again as I haven't seriously touched those yet and play with some of my old Python code snippets I have currently parsing my `upsc `output for a Grafana dashboard.  That's going to be at least a few days before I can report anything back.

I'm thinking if were going to iterate through the list of values pulled from NUT, I will need to template some descriptions based on the measurement type (voltage, current, power, realpower, powerfactor, etc.).

My thought process here is:
1) Define a list of measurement types with their default values for that predefined measurement's `SensorEntityDescription`: (key, name, icon, native_unit_of_measurement, entity_category, entity_registry_enabled_default, etc.)
2) Parse full list of `key: values` from NUT
3) Look through keys starting with `outlet`
4) Look at last part of key for a measurement type in a predefined list (voltage, current, power, powerfactor, etc.)
5) Generate new `SensorEntityDescription` with the default values for that predefined measurement (key, name, icon, native_unit_of_measurement, entity_category, entity_registry_enabled_default, etc.)

Could use that same code for keys that start with `ups`, `input` and `output` as well to get rid of some of the existing definitions as well to ensure they are all being handled the same.